packages/SwingSet/src/vats/vat-admin/vat-vat-admin.js

```diff
@@ -19,20 +26,141 @@ function producePRR() {
   return [promise, { resolve, reject }];
 }
 
-export function buildRootObject(vatPowers) {
+export function buildRootObject(vatPowers, _vatParameters, baggage) {
+  const markerKindHandle = provide(baggage, 'marker', () =>
+    makeKindHandle('marker'),
+  );
+  const makeMarker = defineDurableKind(markerKindHandle, () => ({}), {});
+  const criticalVatKey = provide(baggage, 'criticalVatKey', () => makeMarker());
+
   const { D } = vatPowers;
-  const pending = new Map(); // vatID -> { resolve, reject } for promise
-  const running = new Map(); // vatID -> { resolve, reject } for doneP
-  const meterByID = new Map(); // meterID -> { meter, updater }
+  const pendingVatCreations = new Map(); // vatID -> { resolve, reject } for promise
+
+  let vatAdminDev;
+  let vatAdminVatID;
+
+  const runningVats = new Map(); // vatID -> [doneP, { resolve, reject }]
+  const exitingVats = new Set();
+  function noteRunningVat(vatID) {
+    const prr = producePRR();
+    const [doneP] = prr;
+    doneP.catch(() => {}); // shut up false whine about unhandled rejection
+    runningVats.set(vatID, prr);
+  }
+
+  let quiescentP;
+  let quiescentRR;
+
+  const managedVats = provide(baggage, 'managedVats', () =>
+    makeScalarBigSetStore('managedVats', { durable: true }),
+  );
+  for (const vatID of managedVats.keys()) {
```

Comment:
Perhaps in a different version of JavaScript.  `forEach` is a method on `Array` but set store `.keys()` returns an iterator (and you don't want it to return an array because it might be a store because it contains a jillion elements).